Workiva/dart_dev
# UIP-2338 / CP-2144: Support custom local format task + specifying files via CLI

## Description
- There were several tweaks that needed to be made to support a custom local formatter task in dart_dev.

- https://github.com/Workiva/dart_dev/issues/51:
    > By default, the format task should use the project's configuration for files to format. But, we should also allow running the formatter on a specific file or set of files. This can be done by allowing additional args to the format CLI that override the default configuration.

    I also wanted this behavior for my custom formatter task and needed to make some changes in dart_dev to support it, so I thought it'd only be fair to address that issue while I was at it 😃.


## Solution
- Format task:
    - Split out and export logic for expanding lists of formatted files into `getFilesToFormat`, for use in custom format task
    - Speed up run time by:
    ![image](https://cloud.githubusercontent.com/assets/1713967/26464630/4f7a8634-413d-11e7-9c84-bd16d3e07dde.png)
        - Not resolving absolute paths unnecessarily
        - Not following symlinks
            - __Potential breakage:__ this is a behavioral change, although in master, symlinks are only followed when there are excludes present.
     - Allow specifying specific directories/files via CLI rest args
    - Update formatter API to be able to take in specific files directly
    - Deprecate `FormatConfig.directories`, replacing it with and proxying it to `FormatConfig.paths` for clearer naming (since files can now be specified directly), and deprecate/proxy other `directories` named arguments accordingly
    - Update functional tests.
- Add new `TaskCLI.usage` getter that allows for custom usage strings, to allow for indicating that `format`/`test` package accept rest args
    ```diff
     $ ddev format --help
     Standardized tooling for Dart projects.

    -Usage: pub run dart_dev format [options]
    +Usage: pub run dart_dev format [options] [files or directories...]
    ```

- Add `ExportConfig` task (`ddev export-config`) to allow local tasks to read configuration (necessary since they're launched in separate processes)
    - This is a stopgap solution to support the new custom format task until dart_dev 2.0, which will hopefully have better configuration sharing 
    - `Reporer`: don't colorize empty strings (so that `[38;5;2m[0m` isn't appended to JSON output due to printing empty success result string)
- Allow overriding of built-in tasks (e.g., `format`)
- Update Regex to fix AnalyzerTask functional tests that were failing with the latest analyzer, due to CLI output changes

## Testing:
- Verify that functional tests pass
- Verify `ddev format --help` looks good
- Verify `ddev test --help` looks good

## Areas of Regression
- Format task
- Colorized output
- `--help` flag
